File: version_1/example_1/python/generate_widget.py
#!/usr/bin/python
# import the necessary libraries
import json
import urllib.request
import urllib.error
import string
import sys
# set some important variables
domain = "http://jpapi.codingstudio.org"
api_key = YOUR_API_KEY
api_url = domain+"/v1/people_groups/daily_unreached.json?api_key="+api_key
try:
    # request the API for the Daily Unreached People Group
    request = urllib.request.urlopen(api_url)
except urllib.error.HTTPError as e:
    print('The server couldn\'t fulfill the request.')
    print('Error code: ', e.code)
    exit
except urllib.error.URLError as e:
    print('We failed to reach a server.')
    print('Reason: ', e.reason)
    exit
else:
    # decode the response
    response = request.read().decode("utf8")
    # load the JSON
    data = json.loads(response)
    unreached = data[0]
    # format population to be a comma seperated integer
    unreached['Population'] = format(int(unreached['Population']), ',d')
    # check if percent of Evangelicals is None
    if unreached['PCEvangelical'] is None:
        unreached['PCEvangelical'] = '0'
    # format percent of Evangelicals to percent
    unreached['PCEvangelical'] = float(unreached['PCEvangelical'])
    # get the template file
    index_file = open('templates/index.html').read()
    # initialize a new template
    template = string.Template(index_file)
    # make the substitution
    final_code = template.substitute(unreached)
    # create the widget.html file
    widget_file = open('generated_code/widget.html','w')
    widget_file.write(final_code)

File: version_1/example_1/ruby/generate_widget.rb
# We will use Erubis for the templating
require "erubis"
# We need net/http to handle the request to the API
require "net/http"
# We will need to parse the JSON response
require "json"
# set some important variables
domain = "jpapi.codingstudio.org"
api_key = YOUR_API_KEY
api_path = "/v1/people_groups/daily_unreached.json?api_key=#{api_key}"
# locals = {}

begin
	# Make the request to the Joshua Project API
	response = Net::HTTP.get(domain, api_path)
	# Parse the response
	data = JSON.parse(response)
	unreached = data[0]
rescue Exception => e
	# We had an error
	puts "Unable to get the API data"
	puts e.message
	abort
end
# format the population to a comma seperated value
unreached['Population'] = unreached['Population'].to_s.gsub(/(\d)(?=(\d{3})+$)/,'\1,')
# Lets handle the evangelical number since it can be nil
if unreached['PCEvangelical'].nil?
	unreached['PCEvangelical'] = "0.00"
else
	# format the percent to a floating point (decimal)
	unreached['PCEvangelical'] = '%.2f' % unreached['PCEvangelical']
end
# Generate the template
template_file = File.read("templates/index.html.erb")
template = Erubis::Eruby.new(template_file)
# run the Erubis substitution
widget_code = template.result({unreached: unreached})
# We will write the final HTML file
begin
	# open the final file
	file = File.open("generated_code/widget.html", "w")
	# write the new file
	file.write(widget_code)
rescue IOError => e
	# We had an error
 	puts "Unable to write the HTML file"
	puts e.message
	abort
ensure
	# ensure the file closes if this fails
	file.close unless file == nil
end